kernel/src/main/java/org/kframework/kprove/KProve.java

```diff
@@ -44,8 +46,14 @@ public KProve(KExceptionManager kem, Stopwatch sw, FileUtil files, KPrint kprint
 
     public int run(KProveOptions options, CompiledDefinition compiledDefinition, Backend backend, Function<Module, Rewriter> rewriterGenerator) {
         Tuple2<Definition, Module> compiled = getProofDefinition(options.specFile(files), options.defModule, options.specModule, compiledDefinition, backend, files, kem, sw);
-        Rewriter rewriter = rewriterGenerator.apply(compiled._1().mainModule());
+        Definition def = compiled._1();
+        Rewriter rewriter = rewriterGenerator.apply(def.mainModule());
         Module specModule = compiled._2();
+        Module languageModule = def.getModule("LANGUAGE-PARSING").get();
+        PrettyPrinter prettyPrinter = new PrettyPrinter(kprint, def, languageModule);
+        if (rewriter instanceof WantsPrettyPrinter) {
```

Comment:
this type of mutable state is not very nice and is error-prone. You should be able to rewrite this so that you don't need to call any setter function like this.

Reply:
I'm very well aware of this rule. The clean alternative is to pass `PrettyPrinter` to rewriteGenerator here: https://github.com/kframework/k/blob/8c813df953f27e1b13d57383d1920b4df52946c8/kernel/src/main/java/org/kframework/kprove/KProve.java#L50

or to Rewriter here: https://github.com/kframework/k/blob/8c813df953f27e1b13d57383d1920b4df52946c8/kernel/src/main/java/org/kframework/kprove/KProve.java#L58

Both would affect the interface of all backends and would be considerably more invasive.
This PR is really a compromise with design flaws that already exist in kernel module.

Reply:
Okay here's what we're going to do. We're going to change the signature of the K frontend so that a Rewriter is generated from a Definition instead of a Module. I've checked, and as far as I can tell in all cases we just use the main module of a definition anyway, so it should be an easy change to make, if extensive.

Reply:
makes sense. thx!

Reply:
@dwightguth After a more detailed look at the code, your proposal doesn't seem possible.

RewriterGenerator is shared between kprove and krun, but takes different module as argument:
KProve: https://github.com/kframework/k/blob/cc72c39b37aa4d07e8d7da4c34544d488ed24a3e/kernel/src/main/java/org/kframework/kprove/KProve.java#L47

KRun: https://github.com/kframework/k/blob/91a69ed56a92897677d16bbdb2065841a4446125/kernel/src/main/java/org/kframework/krun/modes/KRunExecutionMode.java#L43

So it's not possible to replace module with definition because you won't know how to get the right module in the right context.

The way I'd do it is to modify the interface of RewriterGenerator to take 2 arguments - module and PrettyPrinter. Then I'd keep `PrettyPrinter` because it nicely encapsulates all that's needed for pretty printing, but remove `WantsPrettyPrinter`.

Does this look good to you?

Reply:
If you look at the implementation of the function, executionModule() is just kompiledDefinition.mainModule(). So it's still the main module of a definition in both cases, which means your concern doesn't really apply. It should still be doable to pass in a Definition unless you find another call site somewhere else that does something other than one of these two things.